min_heap_binary_tree.rb

```diff
@@ -80,4 +101,68 @@ def parent(index)
   def parents_index(index)
     index / 2
   end
+
+  def left_child_index(index)
+    index * 2
+  end
+
+  def right_child_index(index)
+    (index * 2) + 1
+  end
+
+  def index_to_swap_with_two_children(index)
+    if right_child(index).value <= left_child(index).value
+      (2 * index) + 1
+    else
+      2 * index
+    end
+  end
+
+  def only_left_child_smaller?(index)
+    (left_child(index).value < tree[index].value) && (tree[index].value <= right_child(index).value)
+  end
+
+  def only_right_child_smaller?(index)
+    (tree[index].value < left_child(index).value) && (right_child(index).value < tree[index].value)
+  end
+
+  def left_child_larger?(index)
+    left_child(index) != nil && tree[index].value > left_child(index).value
+  end
+
+  def both_children_smaller_swap(replacement_index)
+    temp = index_to_swap_with_two_children(replacement_index)
+    swap! replacement_index, index_to_swap_with_two_children(replacement_index)
+    swap_down(tree[temp])
+  end
+
+  def two_children_swap(replacement_index)
+    if only_left_child_smaller?(replacement_index)
+      swap! replacement_index, left_child_index(replacement_index)
+    elsif only_right_child_smaller?(replacement_index)
+      swap! replacement_index, right_child_index(replacement_index)
+    else
+      both_children_smaller_swap(replacement_index)
+    end
+  end
+
+  def has_two_children?(replacement_index)
+    left_child(replacement_index) != nil && right_child(replacement_index) != nil
+  end
+
+  def swap_down(node)
+    replacement_index = tree.find_index(node)
+
+    if has_two_children?(replacement_index)
```

Comment:
It looks like this is being checked two times in a row.  First on line 26, then again here.

It looks like you can remove this second check, and add in an `elsif` to the after line 26 checking for having only a left child.

So it would be like,

```
if the parent is greater than the replace node, do something
elsif there are two kids, do something,
elsif there is only one kid, do something
```